### PR TITLE
added pulse check to check BE health

### PR DESCRIPTION
**Pull request by shreyakhajanchi**

### CODE REVIEW

File: ui/src/app/services/backend-health/backend-health.service.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    checkBackendHealth() {
+      this.checkHealth().subscribe(
+        (isHealthy) => {
+          if (!isHealthy) {
+            if (this.unHealthyCheckCount ==1) {
+              // Backend is unhealthy, open the dialog and unsubscribe
+              this.openHealthDialog();
+            }
+            this.unHealthyCheckCount=1;
+          }
```
Comment (manitgupta) [inline ui/src/app/services/backend-health/backend-health.service.ts:40]: I think we should have successive health checks failing for a period of X seconds/minutes before showing this to account for possible flakiness. 

Suggest something like 10 consecutive health checks failed so the popup shows up in 50 seconds, or something similar.

File: ui/src/app/services/backend-health/backend-health.service.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        }
+      );
+    }
+
+  openHealthDialog() {
+    let dialogRef = this.dialog.open(InfodialogComponent, {
+      width: '500px',
+      data: {
+        message: 'Please check terminal logs for more details. In case of a crash please file a github issue with all the details.',
```
Comment (manitgupta) [inline ui/src/app/services/backend-health/backend-health.service.ts:49]: Nit: Is it possible to hyperlink `github` here?

File: ui/src/app/services/backend-health/backend-health.service.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    return from(this.fetch.checkBackendHealth()).pipe(
+      map(() => true),
+      catchError(() => {
+        return of(false);
+      })
+    );
+  }
+
+  isHealthy(): boolean {
+    return this.isBackendHealthy;
```
Comment (manitgupta) [inline ui/src/app/services/backend-health/backend-health.service.ts:70]: Am I missing something, I see that `isBackendHealthy` is initialised to `true` and then checked here. Wont `isHealthy()` always return true?

Comment (manitgupta) [reply ui/src/app/services/backend-health/backend-health.service.ts:70]: I think maybe one of `isBackendHealthy` and `unhealthyCheckCount` is needed?

File: ui/src/app/services/backend-health/backend-health.service.ts (hunk 2 above)
Comment (shreyakhajanchi) [reply ui/src/app/services/backend-health/backend-health.service.ts]: Added link

File: ui/src/app/services/backend-health/backend-health.service.ts (hunk 3 above)
Comment (shreyakhajanchi) [reply ui/src/app/services/backend-health/backend-health.service.ts]: Yes this method was not being used, removed

File: ui/src/app/services/backend-health/backend-health.service.ts (hunk 1 above)
Comment (shreyakhajanchi) [reply ui/src/app/services/backend-health/backend-health.service.ts]: added 5 successive checks let me know if you think 10 is ideal number

File: ui/src/app/services/backend-health/backend-health.service.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        this.healthCheckSubscription.unsubscribe();
+      }
+    }
+  
+    checkBackendHealth() {
+      this.checkHealth().subscribe(
+        (isHealthy) => {
+          if (!isHealthy) {
+            if (this.unHealthyCheckCount == 5) {
+              // Backend is unhealthy, open the dialog and unsubscribe
```
Comment (manitgupta) [inline ui/src/app/services/backend-health/backend-health.service.ts:35]: Nit: Declare `5` as `MAX_UNHEALTHY_CHECK_ATTEMPTS`

File: ui/src/app/services/backend-health/backend-health.service.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+          }
+        }
+      );
+    }
+
+  openHealthDialog() {
+    let dialogRef = this.dialog.open(InfodialogComponent, {
+      width: '500px',
+      data: {
+        message: "Please check terminal logs for more details. In case of a crash please file a <a href='https://github.com/GoogleCloudPlatform/spanner-migration-tool/issues' target='_blank' class='a-link'>github</a> issue with all the details.",
```
Comment (manitgupta) [inline ui/src/app/services/backend-health/backend-health.service.ts:50]: "an issue"

File: ui/src/app/services/backend-health/backend-health.service.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        type: 'error',
+        title: 'Backend server unresponsive',
```
Comment (manitgupta) [inline ui/src/app/services/backend-health/backend-health.service.ts:52]: Maybe just write "Spanner migration tool unresponsive". Folks may get confused that backend means a remote service or something.

File: webv2/routes.go [hunk 7]
```diff
@@ -114,6 +114,7 @@ func getRoutes() *mux.Router {
 	router.HandleFunc("/uploadFile", uploadFile).Methods("POST")
 
 	router.HandleFunc("/GetTableWithErrors", getTableWithErrors).Methods("GET")
+	router.HandleFunc("/health", getBackendHealth).Methods("GET")
```
Comment (manitgupta) [inline webv2/routes.go:117]: Nit; generally heartbeat APIs are named `ping`

File: ui/src/app/services/backend-health/backend-health.service.ts (hunk 5 above)
Comment (shreyakhajanchi) [reply ui/src/app/services/backend-health/backend-health.service.ts:50]: its actually "a github issue"

**Review by manitgupta — APPROVED** [review]

**Pull request merged**